machine_learning/dimensionality_reduction.py

```diff
@@ -0,0 +1,199 @@
+        logging.basicConfig(level=logging.ERROR, format="%(message)s", force=True)
+        logging.error("Dataset empty")
+        raise AssertionError
+
+
+def linear_discriminant_analysis(
+    features: np.ndarray, labels: np.ndarray, classes: int, dimensions: int
+) -> np.ndarray:
+    """
+    Linear Discriminant Analysis.
+
+    For more details, see: https://en.wikipedia.org/wiki/Linear_discriminant_analysis.
+    Parameters:
+        * features: the features extracted from the dataset
+        * labels: the class labels of the features
+        * classes: the number of classes present in the dataset
+        * dimensions: to filter the projected data for the desired dimension
+    """
+
+    # Check if the dimension desired is less than the number of classes
+    assert classes > dimensions
+
+    # Check if features have been already loaded
+    if features.any:
+        _, eigenvectors = eigh(
+            covariance_between_classes(features, labels, classes),
+            covariance_within_classes(features, labels, classes),
+        )
+        filtered_eigenvectors = eigenvectors[:, ::-1][:, :dimensions]
+        svd_matrix, _, _ = np.linalg.svd(filtered_eigenvectors)
+        filtered_svd_matrix = svd_matrix[:, 0:dimensions]
+        projected_data = np.dot(filtered_svd_matrix.T, features)
+        logging.info("Linear Discriminant Analysis computed")
+
+        return projected_data
+    else:
+        logging.basicConfig(level=logging.ERROR, format="%(message)s", force=True)
+        logging.error("Dataset empty")
+        raise AssertionError
+
+
+def test_linear_discriminant_analysis() -> None:
+    # Create dummy dataset with 2 classes and 3 features
+    features = np.array([[1, 2, 3, 4, 5], [2, 3, 4, 5, 6], [3, 4, 5, 6, 7]])
+    labels = np.array([0, 0, 0, 1, 1])
+    classes = 2
+    dimensions = 2
+
+    projected_data = linear_discriminant_analysis(features, labels, classes, dimensions)
+
+    # Assert that the shape of the projected data is correct
+    assert projected_data.shape == (dimensions, features.shape[1])
+
+    # Assert that the projected data is a numpy array
+    assert isinstance(projected_data, np.ndarray)
+
+    # Assert that the projected data is not empty
+    assert projected_data.any()
+
+    # Assert that the function raises an AssertionError if dimensions > classes
+    try:
+        projected_data = linear_discriminant_analysis(features, labels, classes, 3)
+    except AssertionError:
+        pass
+    else:
+        raise AssertionError("Did not raise AssertionError for dimensions > classes")
+
+
+def test_principal_component_analysis() -> None:
+    features = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+    dimensions = 2
+    expected_output = np.array([[6.92820323, 8.66025404, 10.39230485], [3.0, 3.0, 3.0]])
+    output = principal_component_analysis(features, dimensions)
+    assert np.allclose(
+        expected_output, output
+    ), f"Expected {expected_output}, but got {output}"
```

Comment:
pytest does `assertion rewriting` so it will automatically show `expected_output` and `output` when tests fail.
https://docs.pytest.org/en/7.1.x/example/reportingdemo.html#tbreportdemo

Reply:
Updated